common/web/types/src/util/compiler-interfaces.ts

```diff
@@ -73,16 +73,58 @@ export type CompilerSchema =
   'kpj';
   // | 'keyman-touch-layout.clean'; TODO this has the wrong name pattern, .spec.json instead of .schema.json
 
+/**
+ * A mapping for common path operations, maps to Node path module. This only
+ * defines the functions we are actually using, so that we can port more easily
+ * between different systems.
+ */
+export interface CompilerPathCallbacks {
+  dirname(name: string): string;
+  extname(name: string): string;
+  basename(name: string, ext?: string): string;
+  isAbsolute(name: string): boolean;
+  join(...paths: string[]): string;
+  normalize(p: string): string;
+}
+
+/**
+ * A mapping for common filesystem operations, maps to Node fs module. This only
+ * defines the functions we are actually using, so that we can port more easily
+ * between different systems.
+ */
+export interface CompilerFileSystemCallbacks {
+  readdirSync(name: string): string[];
+  readFileSync(path: string, options?: { encoding?: null; flag?: string; } | null): Uint8Array;
+  readFileSync(path: string, options: { encoding: string; flag?: string; } | string): string;
+  readFileSync(path: string, options?: { encoding?: string | null; flag?: string; } | string | null): string | Uint8Array;
+
+  existsSync(name: string): boolean;
+}
```

Comment:
I'd like to note that if we're considering reuse of related code within a live browser / hosted on a website, I don't believe that any synchronous functions for this exist in that context - it's all async.

It's possible to make related functionality Promise-compatible without too much effort, at least, which would generally match Node's /`fs`'s `async` method variants.  From there, use of `async` / `await`, where appropriate, can mitigate the impact of corresponding changes.

Reply:
It may be helpful to refer to #8267 - the `feature-esmodule-web-engine` PR where I implemented keyboard-loading modules.  For the DOM, I `Promise`-wrapped a dynamically inserted script in a manner that's pretty compatible with what would be obtained via the asynchronous `fs.readFile`.

Looking back... I actually did use the synchronous `readFileSync`  in the node loader, but still returned a `Promise` at the end anyway.  So... maybe not the most perfect example, but it gets the job done.

Reply:
While there is a pretty modern File System API that's becoming supported, its methods are generally asynchronous too.

See https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API#accessing_files - its example uses are marked with `await`.

Reply:
I suppose it does get us mockability and that it does centralize the areas that will need refactoring when the time comes.  If we're fine with this being _only_ that, then okay.  It's less than what the PR's description claims, though.

> The aim here is to provide a surface _that will hopefully match a future web-based component_ such as path-browserify without significant retooling of the kmc-* modules.

I'm still a little fixated on that part of the description text, admittedly.  In this PR's present form, it **_cannot_** meet that stated goal.

It sounds like the best way forward is to acknowledge this as an exercise left until later, and that significant effort will be required due to the incompatibility of the Node sync function signatures with web-based async-only requirements.

Reply:
I think you are making an incorrect assumption here: that we're going to use the File System API, or fs-browserify or similar. That's not a goal of this PR. No "file" system has been chosen. Note also that path-browserify _is_ synchronous (and was used only as an example, not as a decision).

Any async work may well be moved to another part of this future tool that does not yet exist, so that all "file-like" content is available synchronously for the compiler. At this point, I don't know if that will be the case. But it is premature to build in a design for that unknown because it adds significant complexity -- which we may be able to avoid. At the very least, I don't think we need to make that level of changes right now.

Reply:
> I think you are making an incorrect assumption here: that we're going to use the File System API, or fs-browserify or similar. That's not a goal of this PR. No "file" system has been chosen.

I think you're assuming that there exists a component you could choose for use on a website that would provide synchronous file loading.

The assumptions I'm actually making:
1. That you aren't considering preloading files before the compiler
2. Following from item 1, that _any_ file system you could **_possibly_** choose for use on a website will be incompatible if the files are not preloaded.  I'm not picking any specific one - **it's a blanket statement**, based on design of how JS operates and interacts with files in the browser.  I've looked for ways around this before, but it's _all_ asynchronous.

Admittedly, I wasn't fully aware of assumption 1 being an assumption until I read the following block:

> Any async work may well be moved to another part of this future tool that does not yet exist, so that all "file-like" content is available synchronously for the compiler. 

That'd be the **_one_** way I can see to avoid it - to require that all relevant files have been pre-loaded for the compiler before it starts - say, within a developer-like host engine that took care of all the relevant asynchronous loading in advance.

All of that said...

----

> I don't think we need to make that level of changes right now.

This part is fair.